go/pkg/io/impath/impath.go

```diff
@@ -0,0 +1,152 @@
+// Package impath (immutable path) provides immutable and distinguishable types for absolute and relative paths.
+// This allows for string guarantees at compile time, improves legibility and reduces maintenance burden.
+package impath
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+var (
+	ErrNotAbsolute   = errors.New("impath: path is not absolute")
+	ErrNotRelative   = errors.New("impath: path is not relative")
+	ErrNotDescendant = errors.New("impath: target is not descendant of base")
+)
+
+type path string
+
+type Abs struct {
+	path
+}
+
+type Rel struct {
+	path
+}
+
+func (p path) String() string {
+	return string(p)
+}
+
+func (p path) dir() path {
+	return path(filepath.Dir(string(p)))
+}
+
+func (p path) base() path {
+	return path(filepath.Base(string(p)))
+}
+
+func (p Abs) Dir() Abs {
+	return Abs{path: p.dir()}
+}
+
+func (p Rel) Dir() Rel {
+	return Rel{path: p.dir()}
+}
+
+func (p Abs) Base() Abs {
+	return Abs{path: p.base()}
+}
+
+func (p Rel) Base() Rel {
+	return Rel{path: p.base()}
+}
+
+var (
+	zeroAbs = Abs{}
+	zeroRel = Rel{}
+)
+
+// ToAbs creates a new absolute and clean path from the specified parts.
+//
+// If the specified parts do not join to a valid absolute path, ErrNotAbsolute is returned.
+// If error is not nil, the returned value is not valid.
```

Comment:
The value is usually an interface or a pointer that is conventionally nil when an error is returned. Since this function returns a concrete value, I thought it doesn't hurt to reiterate that it should not be used.